core/server/worker/src/main/java/alluxio/underfs/PagedUfsReader.java

```diff
@@ -0,0 +1,307 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.underfs;
+
+import alluxio.conf.AlluxioConfiguration;
+import alluxio.conf.PropertyKey;
+import alluxio.network.protocol.databuffer.NioDirectBufferPool;
+import alluxio.resource.CloseableResource;
+import alluxio.underfs.options.OpenOptions;
+import alluxio.worker.block.io.BlockReader;
+import alluxio.worker.page.UfsBlockReadOptions;
+
+import com.google.common.base.Preconditions;
+import io.netty.buffer.ByteBuf;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.nio.channels.Channels;
+import java.nio.channels.ClosedChannelException;
+import java.nio.channels.ReadableByteChannel;
+
+/**
+ * Block reader that reads from UFS.
+ */
+public class PagedUfsReader extends BlockReader {
+  private static final ByteBuffer EMPTY_BYTE_BUFFER = ByteBuffer.allocate(0);
+  private final long mPageSize;
+  private final AlluxioConfiguration mConf;
+  private final UfsManager.UfsClient mUfsClient;
+  private final UfsInputStreamCache mUfsInStreamCache;
+  private final FileId mFileId;
+  private final long mFileSize;
+  private final UfsBlockReadOptions mUfsBlockOptions;
+  private final long mInitialOffset;
+  private final ByteBuffer mLastPage;
```

Comment:
It's used to avoid doing small reads that opens and closes a file multiple times. I left a comment in the corresponding integration test https://github.com/Alluxio/alluxio/blob/bd2eadfd31f528883e24d7567dee14a04f323210/tests/src/test/java/alluxio/client/fs/io/FileInStreamIntegrationTest.java#L98